src/engine/server_iv.c

```diff
@@ -987,12 +987,14 @@ iv_op_internal(struct ds_iv_ns *ns, struct ds_iv_key *key_iv,
 				  0, ds_iv_done, &cb_info);
 		break;
 	case IV_UPDATE:
+		D_ASSERT(sync != NULL);
```

Comment:
Fixing checkpatch issues now. For some odd reason checkpatch didn't run on the initial PR but did on my push with the fix to the dfs tests.


The issue is that on line 1034 of engine/server_iv.c it passes sync into iv_op_internal without a null check. Further down that function it checks if sync is not null and if so deferences it and checks something from it. Coverity's issue is that the null check implies that sync can be null but we can pass it into iv_op_internal where its dereferenced without a null check. Looking at the code it looks like it can be null of the op for the iv op is a fetch but if its an update or an invalidate sync is expected to be an actual struct. So my proposed solution is adding 

My initial fix was adding this code to the top of iv_op_internal but then Di suggested the asserts instead.

	if ( (opc == IV_UPDATE || opc == IV_INVALIDATE) && !sync) {
		D_DEBUG(DB_MD, "Attempting IV_UPDATE or IV_INVALIDATE without valid sync type\n");
		return -DER_IO_INVAL;

Unless D_ASSERT compiles out completely on a release build (not sure if it does) then the D_ASSERT will make Coverity happy for everything. I already ran it through with the D_ASSERT and it fixed 2 issues one for each ASSERT.

Reply:
Kind of depends on the assert, no?   If we correctly use asserts such that they only protect against internal coding errors, then I don't see why they can't resolve the issues, particularly ones where coverity can't determine statically whether or not it's possible to hit the issue.

Reply:
Last time I checked (and it was some time ago) if I defined D_ASSERT() to be a no-op our code couldn't even compile.

Personally I'd like to see that fixed and have us run coverity with D_ASSERT compiled out so that we know we're not accidentally (or deliberately in this case) using D_ASSERT() to resolve potential issues.